docs/testing/overview.md

```diff
@@ -1,6 +1,52 @@
+
+Our automated tests fall into three general categories: end-to-end (e2e), integration and unit tests. This document provides comprehensive guidelines on each testing type, emphasizing the importance of choosing the right test for the right scenario and detailing the specific practices for writing and organizing these tests.
+
+## Testing principles
+
+What makes a good test? Here are some general principles.
+
+### 1. Tests should describe expectations
+
+What's the difference between a bug and a feature? A feature is expected, a bug is not.
+
+If it's not clear what the code-under-test is supposed to do, it won't be clear what to do when the test fails either. Clear expectations are essential for an effective test.
+
+### 2. Tests should catch bugs
+
+This is the main purpose of tests. Take care to ensure that each test actually fails when the described expectation is not met.
+
+### 3. Tests should not break often
+
+It's expected that tests will break when we change the behavior or APIs under test. But if a test breaks often from refactors, or from changes to behavior or APIs not directly under test, it's sign that the test may be poorly designed.
+
+Test breakages are expensive. We should do what we can to minimize them.
+
+### 4. Tests should be easy to read
+
+A test isn't effective if we can't understand why it failed. The more readable a test is, the faster we can fix it.
+
+### 5. Tests should be easy to write
+
+Some behaviors are inherently more complex than others, increasing the difficulty of testing them. But in general we want tests to be easy to write, so that we can keep costs low.
+
+If a test is hard to write, consider that as a warning that you might be approaching the problem in a non-optimal way. Resist the urge to power through the work, and instead consider how redesigning the code-under-test or using a different testing strategy might save time.
+
+### 6. Tests should be fast
+
+Making tests faster has a dramatic improvement on productivity, reducing feedback loops.
+
+What we consider "fast" depends on what the test does; what's fast for an end-to-end test might be slow for a unit test. But faster is always better, for all types of tests.
+
+### 7. Tests should have a high return on investment
+
+If a test doesn't align with the previous 6 principles, it's probably doing more harm than good. Consider deleting it.
+
+That doesn't mean we should skip writing tests with it's too difficult. On the contrary; we have high expectations for quality, we don't have the luxury of writing only the easiest most valuable tests. We need to aim for total coverage.
+
+But writing and maintaining bad tests is not the shortest path to complete test coverage. Our time is better spent writing good tests, and making investments that allow us to write more good tests. All code can be designed to be easy to test, and even the most complex test setup/teardown can be reasonably fast and easy to read/write with the right tools.
 
```

Comment:
Would also add that Tests should be isolated:
A good test should not depend on the results or state of another test. Each test should clean up after itself and set up its environment, allowing for parallel or random execution.

Reply:
Great point! I would consider this as covered by "Tests should not break often". The reason I would say that we want tests to be isolated is that non-isolated tests break too often.

Reply:
There is also a tradeoff here to consider I think. For end-to-end tests, it can sometimes be worthwhile to have non-isolated tests for the sake of speeding up test execution time.

We had tests like this in the past and they were miserable to deal with, but with better tooling and better discipline about the expectations before/after each test, I think it can be a good strategy for minimizing test execution time.

Reply:
This is covered already here as well: https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/unit-testing.md#keep-tests-isolated

Reply:
About the tradeoffs for isolated end-to-end tests, this is described here: https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/e2e-testing.md#guidelines-1

I don't completely agree with how it's discussed in that section, but I do agree with the basic point that there are tradeoffs to consider.

Reply:
IIRC, the "test isolation" was already documented here: https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/unit-testing.md#keep-tests-isolated

IDK if we want to make reference to those more "specific" sections of the docs, but aiming to have systemic patterns for our tests and using the same "practices" is definitely valuable too. So if this overview is the "entrypoint" for writing good tests, it could be great to mention them somewhere? (I have no proposal for this though 🙈)

Reply:
ah, I didn't refresh my page before you posted your 2 last comments @Gudahtt sorry 🙈